File: Source/CSBackend/Platform/Android/Video/Base/VideoPlayer.cpp
//
//  VideoPlayer.cpp
//  Chilli Source
//  Created by Ian Copland on 10/08/2012.
//
//  The MIT License (MIT)
//
//  Copyright (c) 2012 Tag Games Limited
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#ifdef CS_TARGETPLATFORM_ANDROID

#include <CSBackend/Platform/Android/Video/Base/VideoPlayer.h>

#include <CSBackend/Platform/Android/Core/JNI/JavaInterfaceManager.h>
#include <CSBackend/Platform/Android/Video/Base/VideoPlayerJavaInterface.h>
#include <ChilliSource/Core/Base/Application.h>
#include <ChilliSource/Core/Delegate/MakeDelegate.h>
#include <ChilliSource/Core/Base/Screen.h>
#include <ChilliSource/Core/Math/MathUtils.h>
#include <ChilliSource/Core/Localisation/LocalisedText.h>
#include <ChilliSource/Core/String/StringUtils.h>
#include <ChilliSource/Core/Threading/TaskScheduler.h>
#include <ChilliSource/Video/Base/Subtitles.h>

namespace CSBackend
{
    namespace Android
    {
    	CS_DEFINE_NAMEDTYPE(VideoPlayer);
        //--------------------------------------------------------------
        //--------------------------------------------------------------
        VideoPlayer::VideoPlayer()
        	: m_currentSubtitleTimeMS(0), m_isPlaying(false)
        {
        }
		//--------------------------------------------------------------
		//--------------------------------------------------------------
		bool VideoPlayer::IsA(CSCore::InterfaceIDType in_interfaceId) const
		{
			return (in_interfaceId == CSVideo::VideoPlayer::InterfaceID || in_interfaceId == VideoPlayer::InterfaceID);
		}
        //--------------------------------------------------------------
        //--------------------------------------------------------------
        void VideoPlayer::Present(CSCore::StorageLocation in_storageLocation, const std::string& in_fileName, VideoCompleteDelegate::Connection&& in_delegateConnection, bool in_dismissWithTap,
        		const CSCore::Colour& in_backgroundColour)
        {
        	CS_ASSERT(m_isPlaying == false, "Cannot present a video while one is already playing.");
        	m_isPlaying = true;

            m_completionDelegateConnection = std::move(in_delegateConnection);

        	//calculate the storage location and full filename.
        	bool isPackage = false;
        	std::string fileName;
        	if (in_storageLocation == CSCore::StorageLocation::k_package)
        	{
        		isPackage = true;
        		fileName = CSCore::Application::Get()->GetFileSystem()->GetAbsolutePathToFile(CSCore::StorageLocation::k_package, in_fileName);
        	}
        	else if (in_storageLocation == CSCore::StorageLocation::k_DLC)
			{
        		if (CSCore::Application::Get()->GetFileSystem()->DoesFileExistInCachedDLC(in_fileName) == true)
        		{
        			isPackage = false;
        			fileName = CSCore::Application::Get()->GetFileSystem()->GetAbsolutePathToStorageLocation(CSCore::StorageLocation::k_DLC) + in_fileName;
        		}
        		else
        		{
        			isPackage = true;
        			fileName = CSCore::Application::Get()->GetFileSystem()->GetAbsolutePathToFile(CSCore::StorageLocation::k_package, CSCore::Application::Get()->GetFileSystem()->GetPackageDLCPath() + in_fileName);
        		}
			}
        	else
        	{
        		isPackage = false;
        		fileName = CSCore::Application::Get()->GetFileSystem()->GetAbsolutePathToStorageLocation(in_storageLocation) + in_fileName;
        	}

        	//start the video
        	m_javaInterface->Present(isPackage, fileName, in_dismissWithTap, in_backgroundColour, CSCore::MakeDelegate(this, &VideoPlayer::OnVideoComplete));
        }
        //--------------------------------------------------------------
		//--------------------------------------------------------------
		void VideoPlayer::PresentWithSubtitles(CSCore::StorageLocation in_storageLocation, const std::string& in_fileName, const CSVideo::SubtitlesCSPtr& in_subtitles, VideoCompleteDelegate::Connection&& in_delegateConnection,
                bool in_dismissWithTap, const CSCore::Colour& in_backgroundColour)
		{
			m_subtitles = in_subtitles;
			m_javaInterface->SetUpdateSubtitlesDelegate(CSCore::MakeDelegate(this, &VideoPlayer::OnUpdateSubtitles));
			Present(in_storageLocation, in_fileName, std::move(in_delegateConnection), in_dismissWithTap, in_backgroundColour);
		}
        //-------------------------------------------------------
        //-------------------------------------------------------
        void VideoPlayer::OnInit()
        {
        	//get the media player java interface or create it if it doesn't yet exist.
			m_javaInterface = JavaInterfaceManager::GetSingletonPtr()->GetJavaInterface<VideoPlayerJavaInterface>();
			if (m_javaInterface == nullptr)
			{
				m_javaInterface = VideoPlayerJavaInterfaceSPtr(new VideoPlayerJavaInterface());
				JavaInterfaceManager::GetSingletonPtr()->AddJavaInterface(m_javaInterface);
			}
        }
        //---------------------------------------------------------------
        //---------------------------------------------------------------
        void VideoPlayer::OnVideoComplete()
        {
        	m_javaInterface->SetUpdateSubtitlesDelegate(nullptr);
        	m_subtitles.reset();
        	m_isPlaying = false;

            if (m_completionDelegateConnection != nullptr)
            {
                auto delegateConnection = std::move(m_completionDelegateConnection);
                m_completionDelegateConnection = nullptr;
                delegateConnection->Call();
            }
        }
		//---------------------------------------------------------------
		//---------------------------------------------------------------
		void VideoPlayer::OnUpdateSubtitles()
		{
			//only update if the position in the video has changed.
			f32 position = m_javaInterface->GetTime();
			TimeIntervalMs currentTimeMS = (TimeIntervalMs)(position * 1000.0f);
			if (m_currentSubtitleTimeMS != currentTimeMS)
			{
				m_currentSubtitleTimeMS = currentTimeMS;

				//get the current subtitles
				auto subtitleArray = m_subtitles->GetSubtitlesAtTime(m_currentSubtitleTimeMS);
				auto localisedText = m_subtitles->GetLocalisedText().get();

				//add any new subtitles
				for (auto it = subtitleArray.begin(); it != subtitleArray.end(); ++it)
				{
					auto mapEntry = m_subtitleMap.find(*it);
					if (mapEntry == m_subtitleMap.end())
					{
						const std::string& text = localisedText->GetText((*it)->m_localisedTextId);
						const CSVideo::Subtitles::Style* style = m_subtitles->GetStyleWithName((*it)->m_styleName);
						s64 subtitleID = m_javaInterface->CreateSubtitle(text, style->m_fontName, style->m_fontSize, CSRendering::StringFromAlignmentAnchor(style->m_alignment), style->m_bounds.vOrigin.x, style->m_bounds.vOrigin.y, style->m_bounds.vSize.x, style->m_bounds.vSize.y);
						m_javaInterface->SetSubtitleColour(subtitleID, 0.0f, 0.0f, 0.0f, 0.0f);
						m_subtitleMap.insert(std::make_pair(*it, subtitleID));
					}
				}

				//update the current text views
				for (auto it = m_subtitleMap.begin(); it != m_subtitleMap.end(); ++it)
				{
					UpdateSubtitle(it->first, it->second, m_currentSubtitleTimeMS);
				}

				//removes any text views that are no longer needed.
				for (auto it = m_subtitlesToRemove.begin(); it != m_subtitlesToRemove.end(); ++it)
				{
					auto mapEntry = m_subtitleMap.find(*it);
					if (mapEntry != m_subtitleMap.end())
					{
						m_javaInterface->RemoveSubtitle(mapEntry->second);
						m_subtitleMap.erase(mapEntry);
					}
				}
				m_subtitlesToRemove.clear();
			}
		}
		//---------------------------------------------------------------
		//---------------------------------------------------------------
		void VideoPlayer::UpdateSubtitle(const CSVideo::Subtitles::Subtitle* in_subtitle, s64 in_subtitleID, TimeIntervalMs in_timeMS)
		{
			const CSVideo::Subtitles::Style* style = m_subtitles->GetStyleWithName(in_subtitle->m_styleName);

			f32 fade = 0.0f;
			s64 relativeTime = ((s64)in_timeMS) - ((s64)in_subtitle->m_startTimeMS);
			s64 displayTime = ((s64)in_subtitle->m_endTimeMS) - ((s64)in_subtitle->m_startTimeMS);

			//subtitle should not be displayed yet so remove
			if (relativeTime < 0)
			{
				m_subtitlesToRemove.push_back(in_subtitle);
			}

			//fading in
			else if (relativeTime < style->m_fadeTimeMS)
			{
				fade = ((f32)relativeTime) / ((f32)style->m_fadeTimeMS);
			}

			//active
			else if (relativeTime < displayTime - style->m_fadeTimeMS)
			{
				fade = 1.0f;
			}

			//fading out
			else if (relativeTime < displayTime)
			{
				fade = 1.0f - (((f32)relativeTime - (displayTime - style->m_fadeTimeMS)) / ((f32)style->m_fadeTimeMS));
			}

			//should no longer be displayed so remove
			else if (relativeTime >= displayTime)
			{
				m_subtitlesToRemove.push_back(in_subtitle);
			}

			m_javaInterface->SetSubtitleColour(in_subtitleID, style->m_colour.r, style->m_colour.g, style->m_colour.b, fade * style->m_colour.a);
		}
        //-------------------------------------------------------
        //-------------------------------------------------------
        void VideoPlayer::OnDestroy()
        {
        	m_javaInterface.reset();
        }
    }
}

#endif

File: Source/CSBackend/Platform/Android/Main/Java/com/chilliworks/chillisource/video/VideoPlayer.java
/**
 * VideoPlayer.java
 * ChilliSource
 * Created by Ian Copland on 10/08/2012.
 * 
 * The MIT License (MIT)
 * 
 * Copyright (c) 2012 Tag Games Limited
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.chilliworks.chillisource.video;

import android.content.Intent;

import com.chilliworks.chillisource.core.CSApplication;
import com.chilliworks.chillisource.core.InterfaceId;
import com.chilliworks.chillisource.core.System;

/**
 * A system which is used to present full screen video from native. This launches a new activity
 * over the OpenGL activity and uses this to play the video in a new view.
 *
 * @author Ian Copland
 */
public class VideoPlayer extends System
{
    public static InterfaceId INTERFACE_ID = new InterfaceId();

	private boolean m_updateSubtitlesEvent = false;

    /**
     * Allows querying of whether or not the system implements the interface described by the
     * given interface id.
     *
     * @author Ian Copland
     *
     * @param in_interfaceId - The interface id to check
     *
     * @return Whether the system implements the given interface
     */
    @Override public boolean isA(InterfaceId in_interfaceId)
    {
        return (in_interfaceId == INTERFACE_ID);
    }
	/**
	 * Presents a video. Videos can be played from file in external storage, from inside the APK
	 * or from inside another file. To play a file from within another (i.e from within a zip other
	 * than the APK) the file offset and length should be set. If playing from a zip then the video
	 * file must not be compressed.
	 *
	 * This is thread-safe.
	 *
	 * @param in_inApk - Whether or not the file is inside the APK.
	 * @param in_filePath - The path to the video file or the file which contains the video file.
	 * @param in_fileOffset - The offset into the file that the video file is located at. This should
	 * be -1 if the file is a video file.
	 * @param in_fileLength - The length of the video file. This should be -1 if the file is a video
	 * file.
	 * @param in_dismissWhenTapped - Whether or not the video activity can be dismissed by a tap.
	 * @param in_hasSubtitles - Whether or not the video has subtitles.
	 * @param in_bgColourR - the red component of the background colour.
	 * @param in_bgColourG - the green component of the background colour.
	 * @param in_bgColourB - the blue component of the background colour.
	 * @param in_bgColourA - the alpha component of the background colour.
	 */
	public void present(boolean in_inApk, String in_filePath, int in_fileOffset, int in_fileLength, boolean in_dismissWhenTapped, boolean in_hasSubtitles,
						float in_bgColourR, float in_bgColourG, float in_bgColourB, float in_bgColourA)
	{
		VideoPlayerActivity.setVideoInfo(in_inApk, in_filePath, in_fileOffset, in_fileLength, in_dismissWhenTapped, in_hasSubtitles, in_bgColourR, in_bgColourG, in_bgColourB, in_bgColourA);
		Intent mediaPlayerIntent = new Intent(CSApplication.get().getActivityContext(), VideoPlayerActivity.class);
		CSApplication.get().getActivityContext().startActivity(mediaPlayerIntent);
	}
	/**
	 * This is thread-safe.
	 *
	 * @author Ian Copland
	 *
	 * @return The current playback position through the video, in seconds.
	 */
	public float getPlaybackPosition()
	{
		VideoPlayerActivity activity = VideoPlayerActivity.GetActivity();
		if (activity != null)
		{
			VideoPlayerView view = activity.GetVideoPlayerView();
			if (view != null)
			{
				return ((float)view.GetTime()) / 1000.0f;
			}
		}
		return 0.0f;
	}
	/**
	 * Creates a new subtitle which will be displayed over the playing video.
	 *
	 * This must be called during the "Update Subtitles" event.
	 *
	 * @author Ian Copland
	 *
	 * @param in_text - The subtitle text.
	 * @param in_fontName - The name of the font to be used.
	 * @param in_fontSize - The font size.
	 * @param in_alignment - The text alignment.
	 * @param in_x - The X position of the rect the text will be located in.
	 * @param in_y - The Y position of the rect the text will be located in.
	 * @param in_width - The width of the rect the text will be located in.
	 * @param in_height - The height of the rect the text will be located in.
	 *
	 * @return The Id of the created subtitle.
	 */
	public long createSubtitle(String in_text, String in_fontName, int in_fontSize, String in_alignment, float in_x, float in_y, float in_width, float in_height)
	{
		assert (m_updateSubtitlesEvent == true) : "Cannot only be called during the Update Subtitles event.";

		VideoPlayerActivity activity = VideoPlayerActivity.GetActivity();
		if (activity != null)
		{
			SubtitlesView view = activity.GetSubtitlesView();
			if (view != null)
			{
				return view.CreateSubtitle(in_text, in_fontName, in_fontSize, in_alignment, in_x, in_y, in_width, in_height);
			}
		}
		return -1;
	}
	/**
	 * Sets the colour of the subtitle with the given Id.
	 *
	 * This must be called during the "Update Subtitles" event.
	 *
	 * @author Ian Copland
	 *
	 * @param in_subtitleId - The Id of the subtitle object.
	 * @param in_red - The red component of the subtitle colour.
	 * @param in_green - The red component of the subtitle colour.
	 * @param in_blue - The red component of the subtitle colour.
	 * @param in_alpha - The red component of the subtitle colour.
	 */
	public void setSubtitleColour(long in_subtitleId, float in_red, float in_green, float in_blue, float in_alpha)
	{
		assert (m_updateSubtitlesEvent == true) : "Cannot only be called during the Update Subtitles event.";

		VideoPlayerActivity activity = VideoPlayerActivity.GetActivity();
		if (activity != null)
		{
			SubtitlesView view = activity.GetSubtitlesView();
			if (view != null)
			{
				view.SetSubtitleColour(in_subtitleId, in_red, in_green, in_blue, in_alpha);
			}
		}
	}
	/**
	 * Removes the subtitle with the given Id.
	 *
	 * This must be called during the "Update Subtitles" event.
	 *
	 * @param in_subtitleId - The Id of the subtitle object which should be removed.
	 */
	public void removeSubtitle(long in_subtitleId)
	{
		assert (m_updateSubtitlesEvent == true) : "Cannot only be called during the Update Subtitles event.";

		VideoPlayerActivity activity = VideoPlayerActivity.GetActivity();
		if (activity != null)
		{
			SubtitlesView view = activity.GetSubtitlesView();
			if (view != null)
			{
				view.RemoveSubtitle(in_subtitleId);
			}
		}
	}
	/**
	 * Called by the Video Player Activity to send the update subtitles event to the native side of
	 * the engine.
	 *
	 * This must be called on the UI thread.
	 *
	 * @author Ian Copland
	 */
	protected void updateSubtitles()
	{
		m_updateSubtitlesEvent = true;
		onUpdateSubtitles();
		m_updateSubtitlesEvent = false;
	}
	/**
	 * Called by the Video Player Activity on completion to signal the native side of the engine.
	 *
	 * This must be called on the UI thread.
	 *
	 * @author Ian Copland
	 */
	protected void completeVideo()
	{
		onVideoComplete();
	}
	/**
	 * Notifies the native side of the engine that it should update subtitles.
	 *
	 * @author Ian Copland
	 */
	private native void onUpdateSubtitles();
	/**
	 * Notifies the native side of the engine that video playback has completed.
	 *
	 * @author Ian Copland
	 */
	private native void onVideoComplete();
}
